### PR TITLE
AVRO-3852: Support Java 21

### PR DESCRIPTION
**Pull request by iemejia**
R: @martin-g

### CODE REVIEW

**Comment by martin-g** [comment]
I guess we need to specify `source` and `target` to `8` at https://github.com/apache/avro/blob/2e03ee9654ca44f16f7c01ad797340b52b56e4e7/lang/java/pom.xml#L116-L118

**Comment by iemejia** [comment]
It should be good now. PTAL again @martin-g

**Review by martin-g — APPROVED** [review]
👍🏻 
One idea for improvement: define properties for the `source` and `target` in the parent pom and use them in the children poms. This way it would be easier to upgrade to a newer version later.

File: lang/java/pom.xml [hunk 1]
```diff
@@ -35,39 +35,40 @@
 
   <properties>
     <main.basedir>${project.parent.basedir}</main.basedir>
+    <java.version>1.8</java.version>
```
Comment (martin-g) [inline lang/java/pom.xml:38]: I think `java.version` is a predefined system property.
You'll need to use a different name here.

File: lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml [hunk 2]
```diff
@@ -137,6 +139,10 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <maven.compiler.source>\${maven.compiler.source}</maven.compiler.source>
```
Comment (martin-g) [inline lang/java/archetypes/avro-service-archetype/src/main/pom/pom.xml:143]:
```suggestion
          <source>\${maven.compiler.source}</source>
```
?

**Comment by iemejia** [comment]
Thanks @martin-g for the advice. I ended up doing a cleaner version now!

**Comment by martin-g** [comment]
@iemejia Does [this](https://github.com/apache/avro/pull/2485/files#diff-dc5428c6ee4d6f032f53378268bf9f79e5812b2f969e837090fd91c96155fe50R143) really work ?

**Comment by iemejia** [comment]
It does, it passed the CI! I just realized that I missed your comment about '<source>'. I will fix that in a subsequent PR where I plan to remove the -ea JDK, once JDK 21 goes GA

**Comment by martin-g** [comment]
The CI does not create a new Maven project from the archetype. That's the reason why it passed.

**Pull request merged**